pyanaconda/modules/storage/iscsi/iscsi.py

```diff
@@ -0,0 +1,240 @@
+        elif mode == "bind":
+            return IscsiInterfacesMode.IFACENAME
+        else:
+            log.error("Unknown iSCSI interface mode %s set by blivet, using UNSET", mode)
+            return IscsiInterfacesMode.UNSET
+
+    def discover_with_task(self, target, credentials, interfaces_mode):
+        """Discover an iSCSI device.
+
+        :param target: the target information
+        :param credentials: the iSCSI credentials
+        :param interfaces_mode: required mode specified by IscsiInterfacesMode string value
+        :return: a DBus path to a task
+        """
+        task = ISCSIDiscoverTask(target, credentials, IscsiInterfacesMode(interfaces_mode))
+        path = self.publish_task(ISCSI.namespace, task, ISCSIDiscoverTaskInterface)
+        return path
+
+    def login_with_task(self, target, credentials, node):
+        """Login into an iSCSI node discovered on a target.
+
+        :param target: the target information
+        :param credentials: the iSCSI credentials
+        :param node: the node information
+        :return: a DBus path to a task
+        """
+        task = ISCSILoginTask(target, credentials, node)
+        path = self.publish_task(ISCSI.namespace, task)
+        return path
+
+    def write_configuration(self, sysroot):
+        """Write the configuration to sysroot."""
+        log.debug("Write iSCSI configuration to %s.", sysroot)
+        iscsi.write(sysroot, None)
+
+    def process_kickstart(self, data):
+        """Process the kickstart data."""
+        if data.iscsiname.iscsiname:
+            self.set_initiator(data.iscsiname.iscsiname)
+        self._iscsi_data = data.iscsi.iscsi
+
+    def setup_kickstart(self, data):
+        """Setup the kickstart data."""
+        data.iscsiname.iscsiname = self.initiator
+        data.iscsi.iscsi = self.generate_iscsi_data(data.IscsiData)
+
+    def generate_iscsi_data(self, iscsi_data_class):
+        """Generate kickstart data based on original kickstart and attached nodes.
+
+        If all nodes for a target were added by a kickstart command, preserve
+        this in generated kickstart (ie do not add particular commands for each
+        discovered node).
+        """
+        iscsi_data_list = self._iscsi_data
+
+        for node in iscsi.active_nodes():
+            if node in iscsi.ibft_nodes:
+                continue
+            iscsi_data = iscsi_data_class()
+            self._setup_iscsi_data_from_node(iscsi_data, node)
+
+            for ks_command_data in iscsi_data_list:
+                # If there is already a (perhaps more general) command
+                # attaching the node, do not add another one
+                if (iscsi_data.ipaddr == ks_command_data.ipaddr and
+                    (not ks_command_data.target or iscsi_data.target == ks_command_data.target) and
+                    iscsi_data.port == ks_command_data.port and
+                    iscsi_data.iface == ks_command_data.iface):
+                    break
```

Comment:
I was thinking, if someone is using
iscsi --ipaddr=10.43.136.51
to attach all nodes on the ip/port, he expects to have the same (attach all nodes on the ip/port)
in generated target rather than the targets that were discovered during installation
iscsi --ipaddr=10.43.136.51 --target=X
iscsi --ipaddr=10.43.136.51 --target=Y

Therefore the care to preserve the input kickstart and to add only targets attached via GUI.

Reply:
Ok.